jupyter_server/services/kernels/kernelmanager.py

```diff
@@ -188,6 +188,17 @@ def cwd_for_path(self, path):
             os_path = os.path.dirname(os_path)
         return os_path
 
+    async def _remove_kernel_when_ready(self, kernel_id, kernel_awaitable):
+        try:
+            await super()._remove_kernel_when_ready(kernel_id, kernel_awaitable)
+        finally:
+            # Unlike its async sibling method in AsyncMappingKernelManager, removing the kernel_id
+            # from the connections dictionary isn't as problematic before the shutdown since the
+            # method is synchronous.  However, we'll keep the relative call orders the same from
+            # a maintenance perspective.
+            self._kernel_connections.pop(kernel_id, None)
+            self._kernel_ports.pop(kernel_id, None)
```

Comment:
Should these occur unconditionally or only on a successful call to the [superclass](https://github.com/jupyter/jupyter_client/blob/c5c0b80084ef3ed01422f2c599276b53c391b774/jupyter_client/multikernelmanager.py#L175-L182)?  Since the KernelManager instance is only removed from the list on a successful `await`, I'm inclined to do the same with connections and ports (and forgo the `try/finally` block).